background/tab-manager.js

```diff
@@ -0,0 +1,43 @@
+/* global navigatorUtil */
+/* exported tabManager */
+'use strict';
+
+const tabManager = (() => {
+  const listeners = [];
+  const cache = new Map();
+  chrome.tabs.onRemoved.addListener(tabId => cache.delete(tabId));
+  chrome.tabs.onReplaced.addListener((added, removed) => cache.delete(removed));
+  navigatorUtil.onUrlChange(({tabId, frameId, url}) => {
+    if (frameId) return;
+    const oldUrl = tabManager.get(tabId, 'url');
+    tabManager.set(tabId, 'url', url);
+    for (const fn of listeners) {
+      try {
+        fn({tabId, url, oldUrl});
+      } catch (err) {
+        console.error(err);
+      }
+    }
+  });
+
+  return {
+    onUpdate(fn) {
```

Comment:
Whenever possible use an inline function because jumping to definition or looking up function source and usages in IDE becomes a real PITA when using separate functions and a link.